File: metroid-map/src/main/java/pl/kwako/metroid_map/metroid_map/KeyPressListener.java
package pl.kwako.metroid_map.metroid_map;

import java.awt.Component;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.util.HashMap;
import java.util.Map;

import static java.awt.event.KeyEvent.*;

public class KeyPressListener extends KeyAdapter {

    private final WindowCoordinateTranslator windowCoordinateTranslator;
    private final Component component;

    /**
     * keeps track of currently active actions
     */
    private int actionMask;

    private static final int MOVE_UP = 0b1;
    private static final int MOVE_DOWN = 0b10;
    private static final int MOVE_LEFT = 0b100;
    private static final int MOVE_RIGHT = 0b1000;
    private static final int ZOOM_IN = 0b10000;
    private static final int ZOOM_OUT = 0b100000;

    private Map<Integer, Integer> keyToActionMapping = new HashMap<>();

    public KeyPressListener(WindowCoordinateTranslator windowCoordinateTranslator, Component component) {
        this.windowCoordinateTranslator = windowCoordinateTranslator;
        this.component = component;

        // TODO: read and write settings using file
        keyToActionMapping.put(VK_UP, MOVE_UP);
        keyToActionMapping.put(VK_W, MOVE_UP);

        keyToActionMapping.put(VK_DOWN, MOVE_DOWN);
        keyToActionMapping.put(VK_S, MOVE_DOWN);

        keyToActionMapping.put(VK_LEFT, MOVE_LEFT);
        keyToActionMapping.put(VK_A, MOVE_LEFT);

        keyToActionMapping.put(VK_RIGHT, MOVE_RIGHT);
        keyToActionMapping.put(VK_D, MOVE_RIGHT);

        keyToActionMapping.put(VK_PAGE_UP, ZOOM_IN);
        keyToActionMapping.put(VK_E, ZOOM_IN);

        keyToActionMapping.put(VK_PAGE_DOWN, ZOOM_OUT);
        keyToActionMapping.put(VK_Q, ZOOM_OUT);
    }

    @Override
    public void keyPressed(KeyEvent e) {

        Integer action = keyToActionMapping.get(e.getKeyCode());

        if (action != null) {
            setAction(action);
        }

        processActions();
    }

    @Override
    public void keyReleased(KeyEvent e) {
        Integer action = keyToActionMapping.get(e.getKeyCode());

        if (action != null) {
            unsetAction(action);
        }
    }

    private void setAction(int action) {
        actionMask |= action;
    }

    private void unsetAction(int action) {
        actionMask &= ~action;
    }

    private void processActions() {
        if (actionSet(MOVE_UP)) {
            windowCoordinateTranslator.moveUp();
        }

        if (actionSet(MOVE_DOWN)) {
            windowCoordinateTranslator.moveDown();
        }

        if (actionSet(MOVE_LEFT)) {
            windowCoordinateTranslator.moveLeft();
        }

        if (actionSet(MOVE_RIGHT)) {
            windowCoordinateTranslator.moveRight();
        }

        if (actionSet(ZOOM_IN)) {
            windowCoordinateTranslator.zoomIn();
        }

        if (actionSet(ZOOM_OUT)) {
            windowCoordinateTranslator.zoomOut();
        }

        component.revalidate();
        component.repaint();
    }

    private boolean actionSet(int action) {
        return (actionMask & action) != 0;
    }
}

File: app/src/main/kotlin/pl/kwako/metroid_map/view/KeyPressListener.kt
package pl.kwako.metroid_map.view

import java.awt.Component
import java.awt.event.KeyAdapter
import java.awt.event.KeyEvent
import java.awt.event.KeyEvent.VK_A
import java.awt.event.KeyEvent.VK_D
import java.awt.event.KeyEvent.VK_DOWN
import java.awt.event.KeyEvent.VK_E
import java.awt.event.KeyEvent.VK_LEFT
import java.awt.event.KeyEvent.VK_PAGE_DOWN
import java.awt.event.KeyEvent.VK_PAGE_UP
import java.awt.event.KeyEvent.VK_Q
import java.awt.event.KeyEvent.VK_RIGHT
import java.awt.event.KeyEvent.VK_S
import java.awt.event.KeyEvent.VK_UP
import java.awt.event.KeyEvent.VK_W
import javax.swing.SwingUtilities

private const val MOVE_UP = 0b1
private const val MOVE_DOWN = 0b10
private const val MOVE_LEFT = 0b100
private const val MOVE_RIGHT = 0b1000
private const val ZOOM_IN = 0b10000
private const val ZOOM_OUT = 0b100000

private val KEY_TO_ACTION_MAPPING = mapOf(
        VK_UP to MOVE_UP,
        VK_W to MOVE_UP,
        VK_DOWN to MOVE_DOWN,
        VK_S to MOVE_DOWN,
        VK_LEFT to MOVE_LEFT,
        VK_A to MOVE_LEFT,
        VK_RIGHT to MOVE_RIGHT,
        VK_D to MOVE_RIGHT,
        VK_PAGE_UP to ZOOM_IN,
        VK_E to ZOOM_IN,
        VK_PAGE_DOWN to ZOOM_OUT,
        VK_Q to ZOOM_OUT
)

private val actionToWindowOperation = mapOf(
        MOVE_UP to WindowCoordinateTranslator::moveUp,
        MOVE_DOWN to WindowCoordinateTranslator::moveDown,
        MOVE_LEFT to WindowCoordinateTranslator::moveLeft,
        MOVE_RIGHT to WindowCoordinateTranslator::moveRight,
        ZOOM_IN to WindowCoordinateTranslator::zoomIn,
        ZOOM_OUT to WindowCoordinateTranslator::zoomOut
)

class KeyPressListener(private val component: Component,
                       private val windowCoordinateTranslator: WindowCoordinateTranslator) : KeyAdapter() {

    private var actionMask: Int = 0

    override fun keyPressed(e: KeyEvent) {
        KEY_TO_ACTION_MAPPING[e.keyCode]?.let {
            setAction(it)
            processActions()
        }
    }

    override fun keyReleased(e: KeyEvent) {
        KEY_TO_ACTION_MAPPING[e.keyCode]?.let {
            unsetAction(it)
            processActions()
        }
    }

    private fun setAction(action: Int) {
        actionMask = actionMask.or(action)
    }

    private fun unsetAction(action: Int) {
        actionMask = actionMask.and(action.inv())
    }

    private fun processActions() {
        actionToWindowOperation.filterKeys { actionSet(it) }
                .forEach { (_, operation) -> operation(windowCoordinateTranslator) }

        SwingUtilities.invokeLater {
            component.revalidate()
            component.repaint()
        }
    }

    private fun actionSet(action: Int): Boolean {
        return actionMask.and(action) != 0
    }
}